InvoiceNumbering/Reactor.fs

```diff
@@ -0,0 +1,14 @@
+module InvoiceNumberingReactor
+
+type Service(numberService: InvoiceNumbering.Service, invoiceService: Invoice.Service) =
+  member _.Handle(streamName, event) =
+    async {
+      match streamName with
+      | FsCodec.StreamName.CategoryAndId(Invoice.Category, Invoice.InvoiceId.Parse invoiceId) ->
+        match Invoice.Events.codec.TryDecode event with
+        | ValueSome(Invoice.Events.InvoiceRaised _) ->
+          let! reservedNumber = numberService.ReserveNext(invoiceId)
```

Comment:
Maybe make the call lazy/async for efficiency under reprocessing? (means TransactAsync vs Transact so maybe not best for a sample tho)

Reply:
This block would become
```
    let resolveNumber () : Async<InvoiceNumber> = numberService.ReserveNext(invoiceId)

    do! invoiceService.EnsureNumber(invoiceId, resolveNumber)
```    
(that resolve is overloaded, but `allocate` or `generate` would not sufficiently emphasize that it might return one from a previous call)
the `decide` function would be supplied the `resolveNumber` argument and would call it iff the state showed that the number was `None`

Reply:
Riight, I see. You're trying to minimise the work done during retries!

The invoice service would receive a function to reserve a number for itself. So if it's already reserved it's only reading one stream instead of two!

Reply:
Yes, one might say its making Decisions for itself ;)

Obviously this is an overkill case in the context of the actual realistic need here.
But this makes way more sense when an indexer reactor is feeding Shipments that have been physically dropped into the shipping container (when the final content that made it based on the initial pick list is known) into batches for financial reporting.

In that context, you run with a DOP of 16 with each `Shipped` event triggering:
1. find the active batch epoch
2. write that start point down as an event on the Shipment
3. ingest it into the batch, walking forward from the start point noted in #2

The "note the starting point and write it down (or just load it if that has already been done)" can be done 16 in parallel (each is an individual roundtrip to a stream)
the next phase, "register them in the batch", is done as a single walk (using an `AsyncBatchingGate`) with the 16 requests merged into one (if you ran them in parallel, they'd continually run into conflicts i.e. WrongExpectedVersionException as they are all trying to merge into the same stream)

But it's definitely a key point that the most general form of such a process is where the overall decision consists of multiple async steps (and the entire chain may need to be rewalked if there is a conflict when you try to store the result as an event). If your only pattern is to feed values into the Service method and/or as a Command, then the decision cannot be dependent on something that's in the aggregate's state. (Of course as usual, none of this potential stuff applies in this instance)

Reply:
I've changed this to conform with your suggestion. I think that's a lot better overall. Thanks